crates/hir_def/src/child_by_source.rs

```diff
@@ -17,145 +18,196 @@ use crate::{
 };
 
 pub trait ChildBySource {
-    fn child_by_source(&self, db: &dyn DefDatabase) -> DynMap {
+    fn child_by_source(&self, db: &dyn DefDatabase, file_id: HirFileId) -> DynMap {
         let mut res = DynMap::default();
-        self.child_by_source_to(db, &mut res);
+        self.child_by_source_to(db, file_id, &mut res);
         res
     }
-    fn child_by_source_to(&self, db: &dyn DefDatabase, map: &mut DynMap);
+    fn child_by_source_to(&self, db: &dyn DefDatabase, file_id: HirFileId, map: &mut DynMap);
 }
 
 impl ChildBySource for TraitId {
-    fn child_by_source_to(&self, db: &dyn DefDatabase, res: &mut DynMap) {
+    fn child_by_source_to(&self, db: &dyn DefDatabase, file_id: HirFileId, res: &mut DynMap) {
         let data = db.trait_data(*self);
-        for (_name, item) in data.items.iter() {
-            match *item {
+        for &(_, item) in data.items.iter() {
+            match item {
                 AssocItemId::FunctionId(func) => {
-                    let src = func.lookup(db).source(db);
-                    res[keys::FUNCTION].insert(src, func)
+                    let loc = func.lookup(db);
+                    if loc.id.file_id() == file_id {
+                        let src = loc.source(db);
+                        res[keys::FUNCTION].insert(src, func)
+                    }
                 }
                 AssocItemId::ConstId(konst) => {
-                    let src = konst.lookup(db).source(db);
-                    res[keys::CONST].insert(src, konst)
+                    let loc = konst.lookup(db);
+                    if loc.id.file_id() == file_id {
+                        let src = loc.source(db);
+                        res[keys::CONST].insert(src, konst)
+                    }
                 }
                 AssocItemId::TypeAliasId(ty) => {
-                    let src = ty.lookup(db).source(db);
-                    res[keys::TYPE_ALIAS].insert(src, ty)
+                    let loc = ty.lookup(db);
+                    if loc.id.file_id() == file_id {
+                        let src = loc.source(db);
+                        res[keys::TYPE_ALIAS].insert(src, ty)
+                    }
```

Comment:
I am afraid of the day we implement [code duplicates search](https://www.jetbrains.com/help/idea/find-and-replace-code-duplicates.html)